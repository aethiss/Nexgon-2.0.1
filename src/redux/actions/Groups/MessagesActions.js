/* eslint-disable no-param-reassign */
import Joi from 'react-native-joi'
import uid from 'uid-safe'

// TODO
// import * as loaderActions from '../Loader/LoaderActions'

export function getAllMessages(groupId) {
  return {
    type: 'GET_ALL_MESSAGES_FROM_GROUP',
    groupId,
  }
}

export function addMessage(message) {
  return {
    type: 'ADD_NEW_MESSAGE_TO_GROUP',
    message,
  }
}

export function removeAllMessages(groupId) {
  return {
    type: 'REMOVE_ALL_MESSAGE_FROM_GROUP',
    groupId,
  }
}

export function createNewGroupMessages(groupId) {
  return {
    type: 'CREATE_NEW_GROUP_MESSAGES',
    groupId,
  }
}

// DEBUG
export function setGroupMessages(store) {
  return {
    type: 'SET_ALL_GROUP_MESSAGES',
    store,
  }
}


// ====== HANDLERS ======= //

export function getAllMessagesHandler(groupId) {
  return (dispatch) => {
    dispatch(getAllMessages(groupId))
  }
}

const messageschema = {
  _type: Joi.string().insensitive(),
  id: Joi.string(),
  createdAt: Joi.string().isoDate(),
  body: {
    text: Joi.string().insensitive(),
  },
  from: {
    _type: Joi.string().insensitive(),
    id: Joi.string(),
    userName: Joi.string().insensitive(),
    userPicture: Joi.string().insensitive(),
  },
  to: {
    _type: 'Group',
    id: Joi.string().alphanum(),
  },
}

export function addMessageHandler(message) {
  return (dispatch) => {
    // dispatch(loaderActions.startLoadingAction())
    message.id = uid.sync(18)
    message.createdAt = new Date().toISOString()
    Joi.validate(message, messageschema, (err) => {
      // dispatch(loaderActions.stopLoadingAction())
      if (err) return (new Error('Error')) // TODO: error logic after validation
      return dispatch(addMessage(message))
    })
  }
}
