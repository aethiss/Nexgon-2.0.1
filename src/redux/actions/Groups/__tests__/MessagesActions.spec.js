/* eslint-disable no-undef */

import * as action from '../MessagesActions'

const snapshot = msgAction => expect(msgAction).toMatchSnapshot()

const dispatch = jest.fn()
const getState = jest.fn()

const newMsg = {
  _type: 'Message',
  // id: 'AAAAAA3333',
  // createdAt: new Date().toISOString(),
  body: {
    text: 'Dite Amici ed entrate !!',
  },
  from: {
    _type: 'User',
    id: 'TTRREE55AAAA',
    userName: 'DbPass',
  },
  to: {
    _type: 'Group',
    id: 'ABCDE123435',
  },
}

describe('@Messages Actions', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('GET all messages action', () => {
    const thunk = action.getAllMessagesHandler('44456DDDCCF')
    // dispatch.mockImplementation(() => Promise.resolve())
    thunk(dispatch, getState)
    expect(dispatch.mock.calls[0][0]).toEqual({
      type: 'GET_ALL_MESSAGES_FROM_GROUP',
      groupId: '44456DDDCCF',
    })
  })
  it('Add new message action', () => {
    const thunk = action.addMessageHandler(newMsg)
    dispatch.mockImplementation(() => Promise.resolve())
    thunk(dispatch, getState)
    expect(dispatch.mock.calls[0][0]).toEqual({
      type: 'ADD_NEW_MESSAGE_TO_GROUP',
      message: newMsg,
    })
  })
  it('expect trow a error if msg schema is not validated', () => {
    const wrongMsg = newMsg
    // eslint-disable-next-line no-underscore-dangle
    wrongMsg._type = 1234
    const thunk = action.addMessageHandler(wrongMsg)
    thunk(dispatch)
  })
  it('REMOVE : all messages from group', () => {
    snapshot(action.removeAllMessages('44455'))
  })
  it('CREATE : new Messages Group', () => {
    snapshot(action.createNewGroupMessages('44455'))
  })
  // DEBUG
  xit('SET : Messages Group', () => {
    expect(action.setGroupMessages([...newMsg])).toMatchSnapshot()
  })
})
