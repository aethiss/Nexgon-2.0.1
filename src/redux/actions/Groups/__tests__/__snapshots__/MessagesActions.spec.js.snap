// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Messages Actions C new Messages Group 1`] = `
Object {
  "groupId": "44455",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions CRE new Messages Group 1`] = `
Object {
  "groupId": "44455",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions CREA new Messages Group 1`] = `
Object {
  "groupId": "44455",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions CREAT new Messages Group 1`] = `
Object {
  "groupId": "44455",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions CREATE  new Messages Group 1`] = `
Object {
  "groupId": "44455",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions CREATE : new Messages Group 1`] = `
Object {
  "groupId": "44455",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions CREATE new Messages Group 1`] = `
Object {
  "groupId": "44455",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions Create new Messages Group 1`] = `
Object {
  "groupId": "55566TTGGF",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions Create new Messages Group 2`] = `
Object {
  "groupId": "44455",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions R all messages from group 1`] = `
Object {
  "groupId": "44455",
  "type": "REMOVE_ALL_MESSAGE_FROM_GROUP",
}
`;

exports[`@Messages Actions REMO all messages from group 1`] = `
Object {
  "groupId": "44455",
  "type": "REMOVE_ALL_MESSAGE_FROM_GROUP",
}
`;

exports[`@Messages Actions REMOV all messages from group 1`] = `
Object {
  "groupId": "44455",
  "type": "REMOVE_ALL_MESSAGE_FROM_GROUP",
}
`;

exports[`@Messages Actions REMOVE : all messages from group 1`] = `
Object {
  "groupId": "44455",
  "type": "REMOVE_ALL_MESSAGE_FROM_GROUP",
}
`;

exports[`@Messages Actions REMOVE all messages from group 1`] = `
Object {
  "groupId": "44455",
  "type": "REMOVE_ALL_MESSAGE_FROM_GROUP",
}
`;

exports[`@Messages Actions Remove all messages from group 1`] = `
Object {
  "groupId": "",
  "type": "CREATE_NEW_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions S Messages Group 1`] = `
Object {
  "store": Object {
    "_type": "Message",
    "body": Object {
      "text": "Porkamadonna !",
    },
    "createdAt": "2017-09-23T15:28:17.358Z",
    "from": Object {
      "_type": "User",
      "id": "TTRREE55AAAA",
      "userName": "DbPass",
      "userPicture": "c://home/ios",
    },
    "id": "GHJJA44443AA654343",
    "to": Object {
      "_type": "Group",
      "id": "ABCDE123435",
    },
  },
  "type": "SET_ALL_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions SET : Messages Group 1`] = `
Object {
  "store": Array [],
  "type": "SET_ALL_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions SET Messages Group 1`] = `
Object {
  "store": Object {
    "_type": "Message",
    "body": Object {
      "text": "Porkamadonna !",
    },
    "createdAt": "2017-09-23T15:28:18.005Z",
    "from": Object {
      "_type": "User",
      "id": "TTRREE55AAAA",
      "userName": "DbPass",
      "userPicture": "c://home/ios",
    },
    "id": "GHJJA44443AA654343",
    "to": Object {
      "_type": "Group",
      "id": "ABCDE123435",
    },
  },
  "type": "SET_ALL_GROUP_MESSAGES",
}
`;

exports[`@Messages Actions SET: Messages Group 1`] = `
Object {
  "store": Object {
    "_type": "Message",
    "body": Object {
      "text": "Porkamadonna !",
    },
    "createdAt": "2017-09-23T15:28:55.729Z",
    "from": Object {
      "_type": "User",
      "id": "TTRREE55AAAA",
      "userName": "DbPass",
      "userPicture": "c://home/ios",
    },
    "id": "GHJJA44443AA654343",
    "to": Object {
      "_type": "Group",
      "id": "ABCDE123435",
    },
  },
  "type": "SET_ALL_GROUP_MESSAGES",
}
`;
